File: new/group_anagrams_sorting.cpp
class Solution {
public:
    vector<vector<string>> groupAnagrams(vector<string>& strs) {
        // create a unordered map to store sortedString: [...list of anagrams for that sorted string]
        // for each string in the given list, sort it, use it as key, and append the current original string
        // to the sorted string, for each sorted string, we have its list of anagrams from original list
        // by the end of the iteration
        // for each pair the map, append the second item to a master list of list of strings and return it. 

        unordered_map<string, vector<string>> map;
        for (const auto &s: strs) { // const means we promise we won't change the string
            string sortedString = s;
            sort(sortedString.begin(), sortedString.end());
            map[sortedString].push_back(s); // we can directly append the values
        }

        vector<vector<string>> master;
        for (auto &kvPair: map) { // each map element is some sort of key value pair of type 'pair'
            master.push_back(kvPair.second); // we can access keys with pair.first, which cannot be modified
            // we can access values with pair.second which can be modified
        }

        return master;
        // time: O(m * nlogn) - m for length of list, n for length of longest string (nlogn for sorting each string)
        // space: O(m * n) - for the map we could have m keys, and could have n list of strings for each key
    }
};

File: new/python/group_anagrams_sorting.py
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        # create a default dictionary for map
        # loop through each string
        # sort each string based on lexical order
        # use sorted string as key and append the current original string in the value
        # for the key which is the sorted version of the current original string
        # loop the map, and get the values, which are list of original strings, grouped
        # together, and return the list
        map = defaultdict(list)  # we have to tell what the default type of values are
        for s in strs:
            # sorted(string) -> list of sorted characters, thus we use ''.join -> string that is sorted character wise
            # we can directly use append(s) because of defaultDict() if it is just dict() we would have to first check if the key exists if it is not, we have to initialize empty array first [] and then later append it.
            map["".join(sorted(s))].append(s)

        # map.values() gives view, not the list those are dictionary items
        # we have to convert them back to list.
        return list(map.values())
        # time: O(m * nlogn) - m for the length of the list, n for length of longest string in a list.
        # space: O(m * n) Each string’s sorted version (key) is length n → up to m unique keys → O(m × n).
        # Each original string (value) is stored once in the map → O(m × n) again.
